extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,204 @@
+    };
+}
+
+#[derive(Debug, Parser, Getters)]
+#[getset(get = "pub")]
+#[clap(version)]
+pub struct Subcommand {
+    /// The tar file image to search and fingerprint JARs in.
+    image_tar_file: PathBuf,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
+struct DiscoveredJar {
+    path: PathBuf,
+    fingerprint: Combined,
+}
+
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+struct OciManifest {
+    #[serde(rename = "Layers")]
+    layers: Vec<PathBuf>,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, TypedBuilder)]
+struct JarAnalysis {
+    // A vector of the Jars discovered during an analysis.
+    discovered_jars: HashMap<PathBuf, Vec<DiscoveredJar>>,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Encountered when we can't open the requisite file.
+    #[error("Open file {0}")]
+    OpenFile(std::io::Error),
+
+    /// Encountered when an archive operation fails.
+    #[error("Tar operation '{0}': {1}")]
+    ReadTar(String, std::io::Error),
+
+    /// Encountered when a manifest file is found but cannot be parsed.
+    #[error("Parsing manifest file {0} {1}")]
+    ParseManifest(PathBuf, serde_json::Error),
+
+    #[error("Reading buffer {0}")]
+    ReadBuffer(std::io::Error),
+
+    #[error("Generating fingerprint {0}")]
+    GenerateFingerprint(fingerprint::Error),
+}
+
+#[tracing::instrument]
+pub fn main(opts: Subcommand) -> Result<(), Report> {
+    let tar_filename = opts.image_tar_file();
+    let jar_analysis = JarAnalysis {
+        discovered_jars: jars_in_container(opts.image_tar_file())
+            .with_context(|| format!("Analyze Jar Archive {:?}", tar_filename))?,
+    };
+
+    serde_json::to_writer(std::io::stdout(), &jar_analysis).context("Serialize Results")
+}
+
+/// Extracts the container (saved via `docker save`) and finds JAR files inside any layer.
+/// For each one found, fingerprints it and reports all those fingerprints along with their
+/// layer and path.
+#[tracing::instrument]
+fn jars_in_container(image_path: &PathBuf) -> Result<HashMap<PathBuf, Vec<DiscoveredJar>>, Error> {
+    // Visit each layer and fingerprint the JARs within.
+    info!("inspecting container");
+    // May have to make two copies of this Archive, idk if it can be iterated twice.
+    let layers = list_container_layers(image_path)?;
+    let mut discoveries = HashMap::new();
+
+    let mut image_archive = unpack(image_path)?;
+    for entry in wrap_tar_err!("Iterate entries", image_archive.entries())? {
+        let entry = wrap_tar_err!("Read entry", entry)?;
+        let path = wrap_tar_err!("Read path", entry.path())?;
+        if !layers.contains(path.as_ref()) {
+            debug!(?path, "skipped: not a layer file");
+            continue;
+        }
+
+        let layer = path.to_path_buf();
+        // Layers should have a form like blob
+        let layer_discoveries = jars_in_layer(entry)?;
+        discoveries.insert(layer, layer_discoveries);
+    }
```

Comment:
[nit, optional] if we used `stable_eyre` for this function, it'd read a lot cleaner:
```suggestion
    let mut image = unpack(image_path)?;
    for entry in image.entries().context("iterate entries")? {
        let entry = entry.context("read entry")?;
        let path = entry.path().context("read path")?;
        if !layers.contains(path.as_ref()) {
            debug!(?path, "skipped: not a layer file");
            continue;
        }

        let layer = path.to_path_buf();
        // Layers should have a form like blob
        let layer_discoveries = jars_in_layer(entry)
            .with_context(|| format!("read layer '{layer}'"))?;
        discoveries.insert(layer, layer_discoveries);
    }
```